// cypress-xray-plugin/src/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ENV_OPENSSL_SECURE_OPTIONS = exports.ENV_OPENSSL_ROOT_CA_PATH = exports.ENV_PLUGIN_DEBUG = exports.ENV_PLUGIN_NORMALIZE_SCREENSHOT_NAMES = exports.ENV_PLUGIN_OVERWRITE_ISSUE_SUMMARY = exports.ENV_CUCUMBER_DOWNLOAD_FEATURES = exports.ENV_CUCUMBER_UPLOAD_FEATURES = exports.ENV_CUCUMBER_FEATURE_FILE_EXTENSION = exports.ENV_XRAY_STEPS_MAX_LENGTH_ACTION = exports.ENV_XRAY_STEPS_UPDATE = exports.ENV_XRAY_TEST_TYPE = exports.ENV_XRAY_STATUS_SKIPPED = exports.ENV_XRAY_STATUS_PENDING = exports.ENV_XRAY_STATUS_FAILED = exports.ENV_XRAY_STATUS_PASSED = exports.ENV_XRAY_UPLOAD_SCREENSHOTS = exports.ENV_XRAY_UPLOAD_RESULTS = exports.ENV_JIRA_TEST_EXECUTION_ISSUE_DESCRIPTION = exports.ENV_JIRA_TEST_EXECUTION_ISSUE_SUMMARY = exports.ENV_JIRA_CREATE_TEST_ISSUES = exports.ENV_JIRA_ATTACH_VIDEOS = exports.ENV_JIRA_TEST_PLAN_ISSUE_KEY = exports.ENV_JIRA_TEST_EXECUTION_ISSUE_KEY = exports.ENV_JIRA_URL = exports.ENV_JIRA_PROJECT_KEY = exports.ENV_JIRA_API_TOKEN = exports.ENV_JIRA_PASSWORD = exports.ENV_JIRA_USERNAME = exports.ENV_XRAY_CLIENT_SECRET = exports.ENV_XRAY_CLIENT_ID = void 0;
// ================================= //
// | Authentication                | //
// ================================= //
// Xray.
exports.ENV_XRAY_CLIENT_ID = "XRAY_CLIENT_ID";
exports.ENV_XRAY_CLIENT_SECRET = "XRAY_CLIENT_SECRET";
// Jira.
exports.ENV_JIRA_USERNAME = "JIRA_USERNAME";
exports.ENV_JIRA_PASSWORD = "JIRA_PASSWORD";
exports.ENV_JIRA_API_TOKEN = "JIRA_API_TOKEN";
// ================================= //
// | Jira Configuration            | //
// ================================= //
exports.ENV_JIRA_PROJECT_KEY = "JIRA_PROJECT_KEY";
exports.ENV_JIRA_URL = "JIRA_URL";
exports.ENV_JIRA_TEST_EXECUTION_ISSUE_KEY = "JIRA_TEST_EXECUTION_ISSUE_KEY";
exports.ENV_JIRA_TEST_PLAN_ISSUE_KEY = "JIRA_TEST_PLAN_ISSUE_KEY";
exports.ENV_JIRA_ATTACH_VIDEOS = "JIRA_ATTACH_VIDEOS";
exports.ENV_JIRA_CREATE_TEST_ISSUES = "JIRA_CREATE_TEST_ISSUES";
exports.ENV_JIRA_TEST_EXECUTION_ISSUE_SUMMARY = "JIRA_TEST_EXECUTION_ISSUE_SUMMARY";
exports.ENV_JIRA_TEST_EXECUTION_ISSUE_DESCRIPTION = "JIRA_TEST_EXECUTION_ISSUE_DESCRIPTION";
// ================================= //
// | Xray Configuration            | //
// ================================= //
exports.ENV_XRAY_UPLOAD_RESULTS = "XRAY_UPLOAD_RESULTS";
exports.ENV_XRAY_UPLOAD_SCREENSHOTS = "XRAY_UPLOAD_SCREENSHOTS";
exports.ENV_XRAY_STATUS_PASSED = "XRAY_STATUS_PASSED";
exports.ENV_XRAY_STATUS_FAILED = "XRAY_STATUS_FAILED";
exports.ENV_XRAY_STATUS_PENDING = "XRAY_STATUS_PENDING";
exports.ENV_XRAY_STATUS_SKIPPED = "XRAY_STATUS_SKIPPED";
exports.ENV_XRAY_TEST_TYPE = "XRAY_TEST_TYPE";
exports.ENV_XRAY_STEPS_UPDATE = "XRAY_STEPS_UPDATE";
exports.ENV_XRAY_STEPS_MAX_LENGTH_ACTION = "XRAY_STEPS_MAX_LENGTH_ACTION";
// ================================= //
// | Cucumber Configuration        | //
// ================================= //
exports.ENV_CUCUMBER_FEATURE_FILE_EXTENSION = "CUCUMBER_FEATURE_FILE_EXTENSION";
exports.ENV_CUCUMBER_UPLOAD_FEATURES = "CUCUMBER_UPLOAD_FEATURES";
exports.ENV_CUCUMBER_DOWNLOAD_FEATURES = "CUCUMBER_DOWNLOAD_FEATURES";
// ================================= //
// | Plugin Configuration          | //
// ================================= //
exports.ENV_PLUGIN_OVERWRITE_ISSUE_SUMMARY = "PLUGIN_OVERWRITE_ISSUE_SUMMARY";
exports.ENV_PLUGIN_NORMALIZE_SCREENSHOT_NAMES = "PLUGIN_NORMALIZE_SCREENSHOT_NAMES";
exports.ENV_PLUGIN_DEBUG = "PLUGIN_DEBUG";
// ================================= //
// | OpenSSL Configuration         | //
// ================================= //
exports.ENV_OPENSSL_ROOT_CA_PATH = "OPENSSL_ROOT_CA_PATH";
exports.ENV_OPENSSL_SECURE_OPTIONS = "OPENSSL_SECURE_OPTIONS";
